crates/core/src/contracts/erc20/ethereum_erc20.rs

```diff
@@ -24,30 +29,57 @@ abigen!(
 );
 
 /// Abstraction for a Kakarot ERC20 contract.
-pub struct EthereumErc20<'a, P> {
+pub struct EthereumErc20<P> {
     pub address: FieldElement,
-    kakarot_contract: &'a KakarotContract<P>,
+    pub provider: Arc<P>,
+    pub kakarot_address: FieldElement,
 }
 
-impl<'a, P: Provider + Send + Sync + 'static> EthereumErc20<'a, P> {
-    pub const fn new(address: FieldElement, kakarot_contract: &'a KakarotContract<P>) -> Self {
-        Self { address, kakarot_contract }
+impl<P: Provider + Send + Sync + 'static> EthereumErc20<P> {
+    pub const fn new(address: FieldElement, provider: Arc<P>, kakarot_address: FieldElement) -> Self {
+        Self { address, provider, kakarot_address }
     }
 
     pub async fn balance_of(self, evm_address: Address, block_id: BlockId) -> Result<U256, EthApiError> {
         // Prepare the calldata for the bytecode function call
         let calldata = IERC20Calls::BalanceOf(BalanceOfCall { account: evm_address }).encode();
-        let calldata = calldata.into_iter().map(FieldElement::from).collect();
+        let calldata: Vec<FieldElement> = calldata.into_iter().map(FieldElement::from).collect();
```

Comment:
is this needed?

Reply:
afaik we need a vec of felts to pass to the contract call

Reply:
ah yeah but do you need to explicitly type it? Compiler can't derive the type?

Reply:
can try